### ISSUE TITLE
moskserver on Docker fail due to OutOfMemoryError: Java heap space

### ISSUE THREAD

**Issue by AnisimovAleksandr** [jamesdbloom/mockserver#310]
It would be great to be able to set JVM-options (like -Xms1024m -Xmx4096m) for docker-image (docker-compose)...

Stacktrace from log:

```
java.lang.RuntimeException: Exception while sending request
at org.mockserver.client.netty.NettyHttpClient.sendRequest(NettyHttpClient.java:90) ~[mockserver-netty-jar-with-dependencies.jar:na]
at org.mockserver.proxy.http.HttpProxyHandler.sendRequest(HttpProxyHandler.java:173) [mockserver-netty-jar-with-dependencies.jar:na]
at org.mockserver.proxy.http.HttpProxyHandler.channelRead0(HttpProxyHandler.java:152) [mockserver-netty-jar-with-dependencies.jar:na]
at org.mockserver.proxy.http.HttpProxyHandler.channelRead0(HttpProxyHandler.java:42) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:307) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:293) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:428) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:243) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:307) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:293) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:307) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:293) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:307) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:293) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:428) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:276) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:263) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:243) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:307) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:293) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1070) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:912) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:369) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:244) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:307) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:293) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:840) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:131) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:511) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:112) [mockserver-netty-jar-with-dependencies.jar:na]
at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137) [mockserver-netty-jar-with-dependencies.jar:na]
at java.lang.Thread.run(Thread.java:745) [na:1.8.0_111]
Caused by: java.util.concurrent.ExecutionException: java.lang.OutOfMemoryError: Java heap space
at com.google.common.util.concurrent.AbstractFuture$Sync.getValue(AbstractFuture.java:299) ~[mockserver-netty-jar-with-dependencies.jar:na]
at com.google.common.util.concurrent.AbstractFuture$Sync.get(AbstractFuture.java:272) ~[mockserver-netty-jar-with-dependencies.jar:na]
at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:96) ~[mockserver-netty-jar-with-dependencies.jar:na]
at org.mockserver.client.netty.NettyHttpClient.sendRequest(NettyHttpClient.java:66) ~[mockserver-netty-jar-with-dependencies.jar:na]
... 38 common frames omitted
Caused by: java.lang.OutOfMemoryError: Java heap space
java.lang.OutOfMemoryError: Java heap space
```

How to overcome this problem? (in the Dockery-image (Dockery-compose) context)

**Comment by enudler** [comment]
We create a new image with the base of the mock server

Like this:

FROM jamesdbloom/mockserver
RUN sed -i "s/runCommand \"java /runCommand \"java -Xmx512m -Xms256m /g" ./run_mockserver.sh

**Comment by AnisimovAleksandr** [comment]
> "We create a new image with the base of the mock server..."

Yes, it was necessary to do like that.

**To jamesdbloom:**
**Memory leak detection!** 
Run tests (using selenium in the docker, 116 tests) for the project (small h5-games). Intercepts requests through a proxy. (Actually, it was on an OutOfMemoryError)
So, the proxy ate 2GB RAM !!!
Here are the results of the study of this problem (in zip 8 files with ObjectsInstanceMap at different stages of memory leaks):
[ObjectsInstanceMap.zip](https://github.com/jamesdbloom/mockserver/files/779468/ObjectsInstanceMap.zip)

**Comment by jamesdbloom** [comment]
Improved Docker Container
====================

I have now added a new switch to the docker container so it is possible to run it as follows:

```bash
docker run -p 1090:1090 jamesdbloom/mockserver /opt/mockserver/run_mockserver.sh -proxyPort 1090 -genericJVMOptions "-Xmx512m -Xms256m"
```

Memory Leak
==========

I don't believe there is a memory leak, although there is no way to be 100% certain of this.

Do you have any measurement for the size of the requests resulting from you 116 tests?  - You would need to consider **both** requests and responses.

Average highly optimised google search request size = 145KB
Other "typical" [site](https://examples.javacodegeeks.com/core-java/junit/junit-setup-teardown-example/) = 2.1MB

If each test causes 10 requests of size 2.1MB, 116 tests would result in 2.4 GB of data in memory ONLY for request data (assuming 100% efficiency in storing request).  This value ignores the size of the 1,160 requests and the size of objects in Java (such a HashMap Entry, String, etc)

If each test causes 100 request of size 145KB, 116 tests would result in 1.7 GB of data in memory ONLY for request data (assuming 100% efficiency in storing request).  This value ignores the size of the 11,600 requests and the size of objects in Java (such a HashMap Entry, String, etc)

Therefore it would be fairly easy for 116 tests to generation several GB of request data (depending on what is being tested).

Testing Approach
=============

I would suggestion either:
1. Launching a new instance of the proxy on a dynamic port for each test.  That way the proxy will only store requests / responses from each test AND tests can run in parallel.  However you will need to inject the proxy port into your system under test at the start of each test.
2. Running tests in series (as done by default in maven, gradle, SBT, iIntelliJ, etc) then clear the proxy at the start of each test run.  **Without clearing the proxy at the start of each test how do you know if the previous test isn't impacting the following test?**

-----

I'm going to close this ticket because I don't believe there is any evidence of an actual issue and the docker container has also been improved.  If you can provide some evidence (i.e. using JProfiler) to show that part of the MockServer Proxy has a memory leak please re-open or raise a new issue.
